src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1332 @@
+    # bin the lulc values such that 
+    # lulc_array[i,j] == sorted_lucodes[lulc_index[i,j]]
+    lulc_index = numpy.digitize(lulc_array, sorted_lucodes, right=True)
+    is_impervious_array[valid_mask] = (
+        impervious_lookup_array[lulc_index][valid_mask])
+    return is_impervious_array 
+
+
+def adjust_op(ratio_array, avg_ratio_array, near_impervious_lulc_array, 
+        near_road_array):
+    """Apply the retention ratio adjustment algorithm to an array of ratios. 
+    This is meant to be used with raster_calculator. Assumes that the nodata
+    value for all four input arrays is the global NODATA.
+
+    Args:
+        ratio_array (numpy.ndarray): 2D array of stormwater retention ratios
+        avg_ratio_array (numpy.ndarray): 2D array of averaged ratios
+        near_impervious_lulc_array (numpy.ndarray): 2D boolean array where 1 
+            means this pixel is near a directly-connected LULC area
+        near_road_array (numpy.ndarray): 2D boolean array where 1 
+            means this pixel is near a road centerline
+        
+    Returns:
+        2D numpy array of adjusted retention ratios. Has the same shape as 
+        ``retention_ratio_array``.
+    """
+    adjusted_ratio_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    adjustment_factor_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (ratio_array != NODATA) &
+        (avg_ratio_array != NODATA) &
+        (near_impervious_lulc_array != NODATA) &
+        (near_road_array != NODATA))
+
+    # adjustment factor:
+    # - 0 if any of the nearby pixels are impervious/connected;
+    # - average of nearby pixels, otherwise
+    is_impervious = (
+        near_impervious_lulc_array[valid_mask] | 
+        near_road_array[valid_mask])
+    adjustment_factor_array[valid_mask] = (avg_ratio_array[valid_mask] * 
+        (1 - is_impervious))
+
+    # equation 2-4: Radj_ij = R_ij + (1 - R_ij) * C_ij
+    adjusted_ratio_array[valid_mask] = (ratio_array[valid_mask] + 
+        (1 - ratio_array[valid_mask]) * adjustment_factor_array[valid_mask])
+    return adjusted_ratio_array
+
+
+def aggregate_results(aoi_path, aggregations, output_path):
+    """Aggregate outputs into regions of interest.
+
+    Args:
+        aoi_path (str): path to vector of polygon(s) to aggregate over
+        aggregations (list[tuple(str,str,str)]): list of tuples describing the 
+            datasets to aggregate. Each tuple has 3 items. The first is the
+            path to a raster to aggregate. The second is the field name for
+            this aggregated data in the output vector. The third is either
+            'mean' or 'sum' indicating the aggregation to perform.
+        output_path (str): path to write out aggregated vector data
+
+    Returns:
+        None
+    """
+
+    if os.path.exists(output_path):
+        os.remove(output_path)
+    aggregate_vector = gdal.OpenEx(output_path, 1)
+    aggregate_layer = aggregate_vector.GetLayer()
```

Comment:
Hmm I think I changed this slightly since you made this comment. It's now `aggregate_vector = gdal.OpenEx(aoi_path, 1)`, and `aoi_path` is an existing vector, mentioned in the docstring: "`this should be a copy that it's okay to modify.`" Because I was already reprojecting the input vector(s) in `execute`, it made sense to me to pass in that reprojected vector to this function. Perhaps this isn't the best structure? I wasn't able to copy exactly from other models that have an aggregation step because they use the shapefile driver and I'm using GPKG (necessary for longer field names due to the user-defined pollutant names). Copying an input shapefile to an output gpkg did not work.